warehouse/oidc/views.py

```diff
@@ -68,81 +96,119 @@ def oidc_audience(request):
     require_csrf=False,
     has_translations=True,
 )
-def mint_token_from_oidc(request):
-    def _invalid(errors):
-        request.response.status = 422
-        return {"message": "Token request failed", "errors": errors}
+@view_config(
+    route_name="oidc.mint_token",
+    require_methods=["POST"],
+    renderer="json",
+    require_csrf=False,
+    has_translations=True,
+)
+def mint_token_from_oidc(request: Request):
+    try:
+        payload = TokenPayload.model_validate_json(request.body)
+        unverified_jwt = payload.token
+    except ValidationError as exc:
+        return _invalid(
+            errors=[{"code": "invalid-payload", "description": str(exc)}],
+            request=request,
+        )
 
-    if request.flags.disallow_oidc(AdminFlagValue.DISALLOW_GITHUB_OIDC):
+    # We currently have an **unverified** JWT. To verify it, we need to
+    # know which OIDC service's keyring to check it against.
+    # To do this, we gingerly peek into the unverified claims and
+    # use the `iss` to key into the right `OIDCPublisherService`.
+    try:
+        unverified_claims = jwt.decode(
+            unverified_jwt, options=dict(verify_signature=False)
+        )
+        unverified_issuer: str = unverified_claims["iss"]
+    except Exception:
+        return _invalid(
+            errors=[{"code": "invalid-payload", "description": "malformed JWT"}],
+            request=request,
+        )
+
+    # Associate the given issuer claim with Warehouse's OIDCPublisherService.
+    service_name = OIDC_ISSUER_SERVICE_NAMES.get(unverified_issuer)
+    if not service_name:
+        return _invalid(
+            errors=[
+                {
+                    "code": "invalid-payload",
+                    "description": "unknown trusted publishing issuer",
+                }
+            ],
+            request=request,
+        )
+
+    if request.flags.disallow_oidc(OIDC_ISSUER_ADMIN_FLAGS[unverified_issuer]):
         return _invalid(
             errors=[
                 {
                     "code": "not-enabled",
-                    "description": (
-                        "GitHub-based trusted publishing functionality not enabled"
-                    ),
+                    "description": "Trusted publishing functionality not enabled",  # noqa
```

Comment:
We're sort of losing some specificity here, at the very least we can use `OIDC_ISSUER_SERVICE_NAMES` to make this more specific.